### PR TITLE
Loss Scaling

### PR DESCRIPTION
**Pull request by anaruse**
This PR adds a loss scaling feature: http://docs.nvidia.com/deeplearning/sdk/mixed-precision-training/index.html#lossscaling.

You can use the loss scaling feature by just setting the scaling factor using a method `set_loss_scaling_factor()` as below.

    from chainer import training
    
    training.set_loss_scaling_factor(1024.0)
    trainer.run()

When the scaling factor is set and data type of parameters is fp16, a fp32 copy of parameters are created automatically and parameters are updated with fp32.

Please note that only ``StandardUpdater`` is supported for now. After some consensus is obtained that this API is OK, I will make the loss scaling available to other Updaters.

### CODE REVIEW

**Comment by beam2d** [comment]
Thank you for the PR.

I think the loss calling should be a feature of backprop computation instead of framework-wide "mode". For example, if the feature can be implemented by adding `loss_scale` option to `Variable.backward` and `chainer.grad`, that's great. We may also want to make `StandardUpdater` support configuring that argument.

IMO, the fp32 update should be treated separately from the loss scaling setting (and thus it's better to split the PR itself as well). I recommend you to implement it as a feature of `UpdateRule` instead of `Parameter`.

**Comment by anaruse** [comment]
I will split this PR to two PRs: one for fp32 update, the other for loss scaling.

**Comment by anaruse** [comment]
I've just submitted a PR for fp32 update as #3708. Please review #3708 first.

**Comment by anaruse** [comment]
The loss scaling feature was re-implemented as a feature of backprop as suggested and the fp32 update feature was removed from this PR.

A `loss_scale` option is added to `Variable.backward` and `chainer.grad`, so that you can use the loss scaling by adding the `loss_scale` option to `Variable.backward` or `chainer.grad` in your code like below.

    loss = lossfunc(y, t)
    loss.backward(loss_scale=1024)

When loss scaling factor is set, the gradient of the loss is scaled up (multiplied by the scaling factor), and the scaling factor is propagated to whole gradients of parameters in a computational graph along the backprop. The gradients of each parameter are scaled down (divided by the scaling factor) just before the parameter is to be updated, so that in principle the same result can be obtained as when loss scaling is not used.

The `loss_scale` option is also added to `StandardUpdater`, so if you use `StandardUpdater`, you can use loss scaling by adding the `loss_scale` option to `StandardUpdater` in your code like below.

    updater = training.StandardUpdater(..., loss_scale=1024)

**Comment by beam2d** [comment]
Sorry for being late. Could you rebase it to the current master?

**Comment by anaruse** [comment]
OK, conflicts to the master branch were resolved. Please check.

File: chainer/function_node.py [hunk 1]
```diff
@@ -875,6 +878,7 @@ def _backprop(outputs, inputs, grad_required, retain_grad, grads):
                 v = node.get_variable_or_none()
                 if v is not None:
                     v.grad_var = g
+                    v.loss_scale = loss_scale
```
Comment (beam2d) [inline chainer/function_node.py]: It's better to use a "private" name, like `v._loss_scale`, and I think it should also be added to the definition of `Variable`. It's enough to add a class attribute `_loss_scale = None`, which is used as the default value.

Comment (anaruse) [reply chainer/function_node.py]: OK, will change the variable name from `loss_scale` to `_loss_scale` and initialize it as you suggested.

**Comment by anaruse** [comment]
The branch was fixed as you suggested. Explanation about `loss_scale` option was also added to related methods.

File: chainer/optimizer.py [hunk 2]
```diff
@@ -581,7 +589,7 @@ def update(self, lossfun=None, *args, **kwds):
                 self.target.cleargrads()
             else:
                 self.target.zerograds()
-            loss.backward()
+            loss.backward(loss_scale=self.loss_scale)
```
Comment (beam2d) [inline chainer/optimizer.py]: I guess this line fails if `set_loss_scale` has not been called. `loss_scale` should have the default value (`None`) and be checked.

Comment (anaruse) [reply chainer/optimizer.py]: Thank you for pointing it out!  will fix to initialize that `loss_scale`.

**Comment by anaruse** [comment]
Added unit tests for loss scaling. Please check.

File: tests/chainer_tests/test_function_node.py [hunk 3]
```diff
@@ -617,10 +634,16 @@ def forward(self):
         self.y = self.x * self.x
 
     def expected_grad(self):
-        return [2 * self.x * self.gy]
+        ret = [2 * self.x * self.gy]
+        if self.loss_scale is not None:
+            ret = [2 * self.x * self.gy * self.loss_scale]
```
Comment (beam2d) [inline tests/chainer_tests/test_function_node.py]: How about
```
grad = 2 * self.x * self.gy
if ...:
    grad *= self.loss_scale
return [grad]
```

File: tests/chainer_tests/test_function_node.py [hunk 4]
```diff
@@ -617,10 +634,16 @@ def forward(self):
...
+        return ret
 
     def expected_double_grad(self):
-        return [2 * self.gy]
+        ret = [2 * self.gy]
```
Comment (beam2d) [inline tests/chainer_tests/test_function_node.py]: ditto

**Comment by beam2d** [comment]
Thank you! LGTM.

**Comment by beam2d** [comment]
Jenkins, test this please.

File: tests/chainer_tests/test_function_node.py [hunk 5]
```diff
@@ -617,10 +634,16 @@ def forward(self):
...
+        grad = 2 * self.x * self.gy
+        if self.loss_scale is not None:
+            grad *= self.loss_scale
+        return [grad]
 
     def expected_double_grad(self):
-        return [2 * self.gy]
+        ggrad = 2 * self.gy
+        if self.loss_scale is not None:
+            ggrad *=  self.loss_scale
```
Comment (beam2d) [inline tests/chainer_tests/test_function_node.py]: Remove one space after `*=`

Comment (anaruse) [reply tests/chainer_tests/test_function_node.py]: Ah, sorry, I forgot to run flake8...

**Comment by beam2d** [comment]
Jenkins, test this please
